dali/python/nvidia/dali/plugin/pytorch/proxy/__init__.py

```diff
@@ -0,0 +1,326 @@
+                "each pipeline input to be mapped by the proxy callable object"
+            )
+        self.num_inputs = len(self.dali_input_names)
+
+        # Multi-process queue used to transfer data from the pytorch workers to the main process
+        self.dali_input_q = mp.Queue()
+        # Multi-process queue used by the main process to consume outputs from the DALI pipeline
+        self.dali_output_q = queue.Queue()
+        # Thread
+        self.thread = None
+        # Cache
+        self.cache_outputs = dict()
+
+    @property
+    def proxy(self):
+        return _DALIProxy(self.dali_input_names, self.dali_input_q)
+
+    def get_outputs(self, pipe_out_ref: DALIPipelineOutputRef):
+        """
+        Gets the pipeline outputs for a specific batch id. It will keep reading data until the
+        right batch is found, caching the results that were not consumed until a future call
+        """
+        req_batch_id = pipe_out_ref.batch_id
+        req_outputs = None
+        # If the data was already read, just return it (and clear the cache entry)
+        if req_batch_id in self.cache_outputs:
+            req_outputs = self.cache_outputs[req_batch_id]
+            del self.cache_outputs[req_batch_id]
+
+        else:
+            curr_batch_id = None
+            # If not the data we are looking for, store it and keep processing until we find it
+            while req_batch_id != curr_batch_id:
+                torch.cuda.nvtx.range_push("dali_output_q.get")
+                curr_batch_id, curr_processed_outputs = self.dali_output_q.get()
+                torch.cuda.nvtx.range_pop()
+                if curr_batch_id == req_batch_id:
+                    req_outputs = curr_processed_outputs
+                else:
+                    self.cache_outputs[curr_batch_id] = curr_processed_outputs
+        # Unpack single element tuples
+        if isinstance(req_outputs, tuple) and len(req_outputs) == 1:
+            req_outputs = req_outputs[0]
+        return req_outputs
+
+    def thread_fn(self):
+        """
+        Asynchronous DALI thread that gets iteration data from the queue and schedules it
+        for execution
+        """
+        self.pipe.build()  # just in case
+
+        while not self.thread_stop_event.is_set():
+            try:
+                torch.cuda.nvtx.range_push("dali_input_q.get")
+                batch_id, inputs = self.dali_input_q.get(timeout=5)
+                torch.cuda.nvtx.range_pop()
+            except mp.TimeoutError:
+                continue
+            except Empty:
+                continue
+
+            torch.cuda.nvtx.range_push(f"schedule iteration {batch_id}")
+            for idx, input_name in enumerate(self.dali_input_names):
+                self.pipe.feed_input(input_name, inputs[idx])
+            self.pipe._run_once()
+            pipe_outputs = self.pipe.outputs()
+
+            # If exec_dynamic, we can avoid copying the data
+            if not self.pipe.exec_dynamic:
```

Comment:
Possibly in a follow-up PR, I would suggest factoring out the tensor conversion logic to a function:
```Python
def to_torch_tensor(dali_tensor, copy):
```
Then we could just do:
```Python
torch_outputs = [to_torch_tensor(pipe_output.as_tensor(), not self pipe.exec_dynamic)
                 for pipe_output in pipe_outputs]
```
This could be used in the regular pytorch iterator.